fix(rpc): Improve input validation and error handling

thevilledev opened the pull request:
Fixes #13067 

The `rpc-server` was vulnerable to Denial of Service attacks via several RPC commands (`SET_TENSOR`, `GRAPH_COMPUTE`, etc.). Malformed messages could trigger failed assertions (e.g., invalid `ggml_type`) or out-of-bounds reads/writes leading to `GGML_ABORT` calls, crashing the server process.

This PR introduces robust input validation and replaces `abort()` calls with graceful error handling:

- **Type Validation:** `deserialize_tensor` now checks if the `tensor->type` is within the valid `GGML_TYPE_COUNT` range *before* calling `ggml_new_tensor_4d`. Returns `nullptr` on invalid type.
- **Bounds Checks:** Replaced `GGML_ABORT` in `set_tensor`, `set_tensor_hash`, and `get_tensor` handlers with error logging and returning `false` when data/offset parameters are out of buffer bounds.
- **Error Propagation:**
    - `create_node` now checks for `nullptr` return values from `deserialize_tensor` and its recursive calls, propagating `nullptr` upwards on failure. Uses `find` instead of `at` for safer map access.
    - `copy_tensor` now checks for `nullptr` from `deserialize_tensor` and sets the response status to failure if deserialization or bounds checks fail.
    - `graph_compute` now checks for `nullptr` return from `create_node` and returns failure status correctly. The final return value now reflects the actual computation status.
    - `RPC_CMD_GET_ALLOC_SIZE` now checks the return value of `server.get_alloc_size` in the RPC server
    loop. If the call fails, return early to close the connection.

thevilledev committed:
fix(rpc): Improve input validation and error handling

The `rpc-server` was vulnerable to Denial of Service attacks via
several RPC commands (`SET_TENSOR`, `GRAPH_COMPUTE`, etc.). Malformed
messages could trigger failed assertions (e.g., invalid `ggml_type`)
or out-of-bounds reads/writes leading to `GGML_ABORT` calls,
crashing the server process.

This PR introduces robust input validation and replaces `abort()`
calls with graceful error handling:

- **Type Validation:** `deserialize_tensor` now checks if the
  `tensor->type` is within the valid `GGML_TYPE_COUNT` range
  *before* calling `ggml_new_tensor_4d`. Returns `nullptr` on
  invalid type.
- **Bounds Checks:** Replaced `GGML_ABORT` in `set_tensor`,
  `set_tensor_hash`, and `get_tensor` handlers with error
  logging and returning `false` when data/offset parameters
  are out of buffer bounds.
- **Size Checks:** Added safe arithmetic checks (for overflow) in
  `graph_compute` when calculating required message sizes based
  on client-provided `n_nodes` and `n_tensors`. Returns early
  if the reported sizes conflict with the actual message size or
  would lead to overflow.
- **Error Propagation:**
    - `create_node` now checks for `nullptr` return values from
      `deserialize_tensor` and its recursive calls, propagating
      `nullptr` upwards on failure. Uses `find` instead of `at`
      for safer map access.
    - `copy_tensor` now checks for `nullptr` from `deserialize_tensor`
      and sets the response status to failure if deserialization
      or bounds checks fail.
    - `graph_compute` now checks for `nullptr` return from
      `create_node` and returns failure status correctly. The final
      return value now reflects the actual computation status.

These changes improve the RPC server's resilience
against malformed client requests, preventing crashes and ensuring
errors are handled more gracefully.

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

rgerganov commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1234,43 +1259,118 @@ ggml_tensor * rpc_server::create_node(uint64_t id,
+    size_t required_size_before_tensors = min_header_size;
+
+    // Check for overflow before adding nodes_array_bytes
+    if (SIZE_MAX - required_size_before_tensors < nodes_array_bytes) {
```
I don't quite understand the new checks which are being added; could you please provide an example where the existing checks fail to detect invalid input?

thevilledev replied:
So from the old implementation:

- Read `n_nodes`
- Calculate offset `O1 = sizeof(uint32_t) + n_nodes * sizeof(uint64_t)`
- Read `n_tensors` from `input.data() + O1`
- Calculate offset `O2 = O1 + sizeof(uint32_t) + n_tensors * sizeof(rpc_tensor)`
- Read tensors array from `input.data() + O1 + sizeof(uint32_t)`

A large `n_nodes` value  combined with a small `input.size()` would read beyond the actual input buffer.

I'll try to summarise the new approach:

- Reads `n_nodes`
- Calculates `nodes_array_bytes = n_nodes * sizeof(uint64_t)`
- Calculates `required_size_before_tensors = min_header_size + nodes_array_bytes + n_tensors_field_size`
  - With overflow checks for the additions, although overflow is _very_ unlikely
- Crucially: Checks `if (input.size() < required_size_before_tensors)` before attempting to read `n_tensors`
  -  This prevents the out-of-bounds read when trying to access `n_tensors` if `n_nodes` was large.
- Reads `n_tensors`
- Calculates `tensors_array_bytes = n_tensors * sizeof(rpc_tensor)`
- Calculates `required_total_size = required_size_before_tensors + tensors_array_bytes`
    - With overflow check
- Checks `if (input.size() < required_total_size)` before attempting to access the tensors data.

This is quite extensive and might be overly cautious, but happy to hear alternatives!

rgerganov replied:
> A large `n_nodes` value combined with a small `input.size()` would read beyond the actual input buffer.

No, it won't. We check for this at line 1265

thevilledev replied:
Apologies, I think I misunderstood. The only difference then left is the overflow checks. While defensive, unlikely to overflow on a 64-bit system. What do you think, are these still worth having?

rgerganov replied:
The memory is already allocated in the `input` vector, we are just parsing here. I believe we already have the necessary checks.

thevilledev replied:
Alright, reverted to the original approach

rgerganov commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1286,12 +1386,20 @@ bool rpc_server::graph_compute(const std::vector<uint8_t> & input, rpc_msg_graph
     response.result = status;
-    return true;
+    // Return true only if computation succeeded
+    return status == GGML_STATUS_SUCCESS;
```
The result of the computation is set in the `response`; here we return `true` because the input is valid

thevilledev replied:
Fair point, will change it back to the original

thevilledev committed:
refactor(rpc): address pr comments

removed comments and unnecessary returns

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

thevilledev committed:
refactor(rpc): ambiguous nullptr from create_node

rpc_server::create_node could previously return nullptr if the input ID
was 0 (valid) or if an internal error (deserialization, recursion
failure) occurred (invalid). This ambiguity made error handling
difficult for the caller (`graph_compute`).

This commit clarifies the meaning of nullptr:
- `graph_compute` now checks if the input 'id' was non-zero when
  `create_node` returns nullptr, correctly identifying failures
  versus intentional null links.
- `create_node` avoids recursive calls for zero IDs and propagates
  nullptr unambiguously on failure during recursion.

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

thevilledev commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1234,43 +1259,118 @@ ggml_tensor * rpc_server::create_node(uint64_t id,
+    auto it_ptr = tensor_ptrs.find(id);
+    if (it_ptr == tensor_ptrs.end()) {
+        GGML_LOG_ERROR("[%s] tensor id %" PRIu64 " not found in provided tensors\n", __func__, id);
+        return nullptr;
```
Fair point! I have an alternative in mind (which I'll push in a moment):

- `graph_compute` checks if the input 'id' was non-zero when `create_node` returns nullptr. Hopefully identifies failures versus intentional null links.
- `create_node` avoids recursive calls for zero IDs and propagates nullptr unambiguously on failure during recursion.

rgerganov replied:
you should also remove the zero check at the beginning of the function

thevilledev replied:
Done, removed noisy logging while at it

thevilledev commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1034,7 +1047,10 @@ bool rpc_server::set_tensor(const std::vector<uint8_t> & input) {
 
         if (in_tensor->data + offset < p0 || in_tensor->data + offset >= p1 || size > (p1 - in_tensor->data - offset)) {
-            GGML_ABORT("[%s] tensor->data out of bounds\n", __func__);
+            // Replace GGML_ABORT with logging and error return
```
Will do!

thevilledev commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1109,7 +1125,11 @@ bool rpc_server::set_tensor_hash(const std::vector<uint8_t> & input, rpc_msg_set
 
         if (in_tensor->data + offset < p0 || in_tensor->data + offset >= p1 || size > (p1 - in_tensor->data - offset)) {
-            GGML_ABORT("[%s] tensor->data out of bounds\n", __func__);
+            // Replace GGML_ABORT with logging and error return
```
Will do!

thevilledev commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1109,7 +1125,11 @@ bool rpc_server::set_tensor_hash(const std::vector<uint8_t> & input, rpc_msg_set
+            // Replace GGML_ABORT with logging and error return
+            GGML_LOG_ERROR("[%s] tensor data region (data=0x%" PRIx64 ", offset=%" PRIu64 ", size=%zu, hash=0x%" PRIx64 ") out of buffer bounds [0x%zx, 0x%zx)\n",
+                           __func__, in_tensor->data, offset, size, *hash, p0, p1);
+            response.result = 0;
```
I missed this, will remove!

thevilledev commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1174,7 +1194,10 @@ bool rpc_server::get_tensor(const rpc_msg_get_tensor_req & request, std::vector<
             request.tensor.data + request.offset >= p1 ||
             request.size > (p1 - request.tensor.data - request.offset)) {
-                GGML_ABORT("[%s] tensor->data out of bounds\n", __func__);
+                // Replace GGML_ABORT with logging and error return
```
Will do!

thevilledev commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1197,6 +1220,7 @@ bool rpc_server::copy_tensor(const rpc_msg_copy_tensor_req & request, rpc_msg_co
     ggml_tensor * dst = deserialize_tensor(ctx, &request.dst);
     if (src == nullptr || dst == nullptr) {
         GGML_LOG_ERROR("[%s] error deserializing tensors\n", __func__);
+        response.result = 0;
```
Will do!

thevilledev commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1214,6 +1238,7 @@ bool rpc_server::copy_tensor(const rpc_msg_copy_tensor_req & request, rpc_msg_co
                          dst_data + src_size,
                          dst_base,
                          dst_base + dst_buf_sz);
+        response.result = 0;
```
Will do!

thevilledev commented:
> on my opinion it may affects to perfomance, may be use feature flag (via cmake)?

I believe it would be interesting to see what the performance impact of this change is. I'm new to the project so pointers welcome if there's a test suite available which would show that.

Slightly off-topic but related: I think there's plenty of opportunities for similar improvements in the RPC server. From invalid tensor operations to crashing via deep recursion in `create_node` which I would like to also fix. I'd like to work on those one change at a time though.

slaren commented:
`RPC_CMD_GET_ALLOC_SIZE` does not check for errors, and if the call to `get_alloc_size` fails it will leave the client connected in a bad state:
https://github.com/ggml-org/llama.cpp/blob/604f0a0045ecb1988a32dbdad7ba4984df9c7ecd/ggml/src/ggml-rpc/ggml-rpc.cpp#L1470

thevilledev committed:
refactor(rpc): initial zero check in create_node

The caller (`graph_compute`) already checks `id != 0` when handling
a `nullptr` return from `create_node`, correctly distinguishing
intentional null links from actual errors. This makes the initial
`if (id == 0)` check redundant.

Also removes the log message when a tensor ID is not found in the
provided map which was added in this branch.

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

thevilledev committed:
fix(rpc): Handle get_alloc_size failure in server

Check the return value of `server.get_alloc_size` in the RPC server
loop. If the call fails, return early to close the connection.

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

thevilledev commented:
> The best investment of efforts in this direction would be creating a script/job for coverage guided fuzzing.

Sounds good, I can look into that after this PR 👍

thevilledev committed:
refactor(rpc): input size validation in graph_compute

Removes detailed, step-by-step size calculations and overflow
checks in favor of simpler direct comparisons, assuming 64-bit
overflow is unlikely.

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

rgerganov commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -1297,6 +1345,15 @@ bool rpc_server::graph_compute(const std::vector<uint8_t> & input, rpc_msg_graph
+        // If id was non-zero and create_node returned nullptr, it indicates a deserialization error.
+        if (graph->nodes[i] == nullptr && id != 0) {
+            GGML_LOG_ERROR("[%s] failed to create graph node %d (id=%" PRId64 ")\n", __func__, i, id);
+            response.result = GGML_STATUS_FAILED;
```
this is not needed

thevilledev replied:
I have a very specific test payload which hits this issue. It uses deeply nested graphs. If I remove the check then the chosen backend gets a null pointer and crashes. I've only tested this on the Metal backend though. Basically this check avoids a secondary crash later in the process.

Here's the backtrace:

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x98)
    frame #0: 0x0000000100218524 libggml-metal.dylib`ggml_metal_encode_node + 96
libggml-metal.dylib`ggml_metal_encode_node:
->  0x100218524 <+96>:  ldp    x26, x21, [x0, #0x98]
    0x100218528 <+100>: ldr    x19, [x0, #0xa8]
    0x10021852c <+104>: bl     0x10021f834    ; symbol stub for: ggml_is_empty
    0x100218530 <+108>: tbnz   w0, #0x0, 0x100218548 ; <+132>
Target 0: (rpc-server) stopped.
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x98)
  * frame #0: 0x0000000100218524 libggml-metal.dylib`ggml_metal_encode_node + 96
    frame #1: 0x0000000100218458 libggml-metal.dylib`__ggml_backend_metal_set_n_cb_block_invoke + 576
    frame #2: 0x0000000100217f04 libggml-metal.dylib`ggml_backend_metal_graph_compute + 368
    frame #3: 0x0000000100344a20 libggml-base.dylib`ggml_backend_graph_compute + 24
    frame #4: 0x0000000100064858 libggml-rpc.dylib`rpc_server::graph_compute(std::__1::vector<unsigned char, std::__1::allocator<unsigned char>> const&, rpc_msg_graph_compute_rsp&) + 380
    frame #5: 0x0000000100065100 libggml-rpc.dylib`ggml_backend_rpc_start_server + 1648
    frame #6: 0x0000000100003334 rpc-server`main + 3300
    frame #7: 0x0000000198dfeb4c dyld`start + 6000
```

rgerganov replied:
I meant that only `response.result = GGML_STATUS_FAILED` is not needed, not the entire check

thevilledev committed:
refactor(rpc): remove extra status code setting

Removes the explicit setting of `response.result = GGML_STATUS_FAILED`
when `create_node` returns `nullptr` within `graph_compute`.
Primary signal is the `false` return value in case of failure.

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

thevilledev committed:
refactor(rpc): remove redundant check for tensor->type

Breaks CI on ubuntu-cpu-make. Tensor type is uint32_t, thus
the check is not needed.

Signed-off-by: Ville Vesilehto <ville@vesilehto.fi>

rgerganov reviewed (approved):
This looks fine to me now, thanks for addressing my comments. I also did some tests and didn't find performance or functional regressions.

@slaren could you please review as well?

slaren commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -982,8 +982,21 @@ bool rpc_server::buffer_clear(const rpc_msg_buffer_clear_req & request) {
     ggml_tensor * result = ggml_new_tensor_4d(ctx, (ggml_type) tensor->type,
         tensor->ne[0], tensor->ne[1], tensor->ne[2], tensor->ne[3]);
+
+    // ggml_new_tensor_4d might fail if dimensions are invalid, although less likely to crash than invalid type
```
If `ggml_new_tensor` fails it will crash, it will not return `NULL`. The check is still good for future-proofing, but the comment is misleading.
